gdal/frmts/pdf/pdfcreatecopy.cpp

```diff
@@ -2266,6 +2380,7 @@ int GDALPDFWriter::WriteOGRFeature(GDALPDFLayerDesc& osVectorDesc,
         OGRStyleToolH hTool = OGR_SM_GetPart(hSM, iPart, nullptr);
         if (hTool)
         {
+            OGR_ST_SetUnit( hTool, OGRSTUMM, 1000.0 ); // TODO no idea why this is necessary, but everything is bad and sad without it!
```

Comment:
Have a look at  OGRStyleTool::ComputeWithUnit(). This means that you set the output unit to millimeters,  and for input tools expressed in ground units, you apply a scale of 1000, which means the value in ground unit is divided by 1000. So the effect of this is to return the value unmodified for tools in ground units. I'm not completely sure this is really what should be done. Perhaps the DPI and/or cartographic resolution should be taken into account.  Or perhaps the units of the input style string aren't correctly set